File: web/src/api/authClient.test.ts
import { describe, expect, it, vi } from "vitest"

import { AuthClient } from "@/api/authClient"

describe("AuthClient", () => {
  it("POSTs request-code and returns the body", async () => {
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(
        JSON.stringify({
          email: "parent@example.com",
          expiresInSeconds: 600,
          devCode: "123456",
        }),
        { status: 202, headers: { "Content-Type": "application/json" } },
      ),
    )

    const client = new AuthClient("http://localhost:8080", fetchFn)
    const result = await client.requestCode("parent@example.com")

    expect(result.devCode).toBe("123456")
    expect(fetchFn).toHaveBeenCalledOnce()
    const [url, init] = fetchFn.mock.calls[0] as [string, RequestInit]
    expect(url).toBe("http://localhost:8080/api/auth/request-code")
    expect(init.method).toBe("POST")
  })

  it("verifies a code and returns a Bearer session", async () => {
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(
        JSON.stringify({
          accessToken: "tok",
          tokenType: "Bearer",
          adult: { id: "1", email: "parent@example.com", displayName: null },
        }),
        { status: 200, headers: { "Content-Type": "application/json" } },
      ),
    )

    const client = new AuthClient("http://localhost:8080", fetchFn)
    const session = await client.verifyCode("parent@example.com", "123456")

    expect(session.accessToken).toBe("tok")
    expect(session.adult.email).toBe("parent@example.com")
  })

  it("sends Bearer on me and logout", async () => {
    const fetchFn = vi
      .fn()
      .mockResolvedValueOnce(
        new Response(
          JSON.stringify({ id: "1", email: "parent@example.com", displayName: null }),
          { status: 200, headers: { "Content-Type": "application/json" } },
        ),
      )
      .mockResolvedValueOnce(new Response(null, { status: 204 }))

    const client = new AuthClient("http://localhost:8080", fetchFn)
    await client.getMe("tok")
    await client.logout("tok")

    expect(fetchFn.mock.calls[0]?.[1]).toMatchObject({
      headers: { Authorization: "Bearer tok" },
    })
    expect(fetchFn.mock.calls[1]?.[1]).toMatchObject({
      method: "POST",
      headers: { Authorization: "Bearer tok" },
    })
  })

  it("throws with server message when request fails", async () => {
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(JSON.stringify({ message: "Too many code requests" }), {
        status: 429,
        headers: { "Content-Type": "application/json" },
      }),
    )
    const client = new AuthClient("http://localhost:8080", fetchFn)

    await expect(client.requestCode("parent@example.com")).rejects.toThrow(
      "Too many code requests",
    )
  })
})

File: mobile/sharedLogic/src/commonTest/kotlin/org/example/project/AuthClientTest.kt
package org.example.project

import io.ktor.client.HttpClient
import io.ktor.client.engine.mock.MockEngine
import io.ktor.client.engine.mock.respond
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpMethod
import io.ktor.http.HttpStatusCode
import io.ktor.http.headersOf
import io.ktor.serialization.kotlinx.json.json
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.json.Json

class AuthClientTest {

    @Test
    fun requestVerifyMeAndLogoutHappyPath() =
        runTest {
            val mockEngine =
                MockEngine { request ->
                    when {
                        request.url.encodedPath == "/api/auth/request-code" &&
                            request.method == HttpMethod.Post ->
                            respond(
                                content =
                                    """{"email":"parent@example.com","expiresInSeconds":600,"devCode":"123456"}""",
                                status = HttpStatusCode.Accepted,
                                headers = headersOf(HttpHeaders.ContentType, "application/json"),
                            )
                        request.url.encodedPath == "/api/auth/verify-code" &&
                            request.method == HttpMethod.Post ->
                            respond(
                                content =
                                    """{"accessToken":"tok","tokenType":"Bearer","adult":{"id":"1","email":"parent@example.com","displayName":null}}""",
                                status = HttpStatusCode.OK,
                                headers = headersOf(HttpHeaders.ContentType, "application/json"),
                            )
                        request.url.encodedPath == "/api/auth/me" -> {
                            assertEquals("Bearer tok", request.headers[HttpHeaders.Authorization])
                            respond(
                                content =
                                    """{"id":"1","email":"parent@example.com","displayName":null}""",
                                status = HttpStatusCode.OK,
                                headers = headersOf(HttpHeaders.ContentType, "application/json"),
                            )
                        }
                        request.url.encodedPath == "/api/auth/logout" &&
                            request.method == HttpMethod.Post ->
                            respond(
                                content = "",
                                status = HttpStatusCode.NoContent,
                            )
                        else -> error("Unexpected ${request.method} ${request.url.encodedPath}")
                    }
                }

            val httpClient =
                HttpClient(mockEngine) {
                    install(ContentNegotiation) {
                        json(Json { ignoreUnknownKeys = true })
                    }
                }
            val client = AuthClient("http://localhost:8080", httpClient)

            val requested = client.requestCode("parent@example.com")
            assertEquals("123456", requested.devCode)

            val session = client.verifyCode("parent@example.com", "123456")
            assertEquals("tok", session.accessToken)

            assertEquals("parent@example.com", client.getMe("tok").email)
            client.logout("tok")
        }

    @Test
    fun requestCodeSurfacesServerErrorMessage() =
        runTest {
            val mockEngine =
                MockEngine {
                    respond(
                        content = """{"message":"Too many code requests"}""",
                        status = HttpStatusCode.TooManyRequests,
                        headers = headersOf(HttpHeaders.ContentType, "application/json"),
                    )
                }
            val httpClient =
                HttpClient(mockEngine) {
                    install(ContentNegotiation) {
                        json(Json { ignoreUnknownKeys = true })
                    }
                }
            val client = AuthClient("http://localhost:8080", httpClient)

            val error =
                assertFailsWith<AuthApiException> {
                    client.requestCode("parent@example.com")
                }
            assertEquals("Too many code requests", error.message)
            assertFalse(error.unreachable)
        }

    @Test
    fun connectFailureIsFlaggedUnreachable() =
        runTest {
            val mockEngine = MockEngine { throw RuntimeException("Failed to connect to /127.0.0.1:8080") }
            val httpClient =
                HttpClient(mockEngine) {
                    install(ContentNegotiation) {
                        json(Json { ignoreUnknownKeys = true })
                    }
                }
            val client = AuthClient("http://127.0.0.1:8080", httpClient)

            val error = assertFailsWith<AuthApiException> { client.getMe("tok") }
            assertTrue(error.unreachable)
            assertTrue(error.message?.contains("Cannot reach http://127.0.0.1:8080") == true)
        }
}

class AuthSessionTest {

    @Test
    fun verifyPersistsTokenAndLogoutClearsIt() =
        runTest {
            val mockEngine =
                MockEngine { request ->
                    when (request.url.encodedPath) {
                        "/api/auth/verify-code" ->
                            respond(
                                content =
                                    """{"accessToken":"tok","tokenType":"Bearer","adult":{"id":"1","email":"a@b.com"}}""",
                                status = HttpStatusCode.OK,
                                headers = headersOf(HttpHeaders.ContentType, "application/json"),
                            )
                        "/api/auth/logout" ->
                            respond(content = "", status = HttpStatusCode.NoContent)
                        else -> error("Unexpected ${request.url.encodedPath}")
                    }
                }
            val httpClient =
                HttpClient(mockEngine) {
                    install(ContentNegotiation) {
                        json(Json { ignoreUnknownKeys = true })
                    }
                }
            val store = InMemorySecureTokenStore()
            val session =
                AuthSession(
                    client = AuthClient("http://localhost:8080", httpClient),
                    tokenStore = store,
                )

            session.verifyCode("a@b.com", "123456")
            assertTrue(session.isSignedIn())
            assertEquals("tok", store.loadAccessToken())

            session.logout()
            assertNull(store.loadAccessToken())
        }

    @Test
    fun logoutClearsTokenEvenWhenServerCallFails() =
        runTest {
            val mockEngine =
                MockEngine { request ->
                    when (request.url.encodedPath) {
                        "/api/auth/logout" -> throw RuntimeException("Failed to connect")
                        else -> error("Unexpected ${request.url.encodedPath}")
                    }
                }
            val httpClient =
                HttpClient(mockEngine) {
                    install(ContentNegotiation) {
                        json(Json { ignoreUnknownKeys = true })
                    }
                }
            val store = InMemorySecureTokenStore().also { it.saveAccessToken("tok") }
            val session =
                AuthSession(
                    client = AuthClient("http://localhost:8080", httpClient),
                    tokenStore = store,
                )

            val error = assertFailsWith<Throwable> { session.logout() }
            assertTrue(error.message?.contains("Failed to connect") == true)
            assertNull(store.loadAccessToken())
            assertFalse(session.isSignedIn())
        }

    @Test
    fun clearLocalSessionDropsTokenWithoutCallingServer() {
        val mockEngine = MockEngine { error("clearLocalSession must not call the server") }
        val httpClient = HttpClient(mockEngine)
        val store = InMemorySecureTokenStore().also { it.saveAccessToken("tok") }
        val session =
            AuthSession(
                client = AuthClient("http://localhost:8080", httpClient),
                tokenStore = store,
            )

        session.clearLocalSession()

        assertNull(store.loadAccessToken())
        assertFalse(session.isSignedIn())
        assertTrue(mockEngine.requestHistory.isEmpty())
    }
}

class InMemorySecureTokenStoreTest {

    @Test
    fun storesAndClearsToken() {
        val store = InMemorySecureTokenStore()
        store.saveAccessToken("abc")
        assertEquals("abc", store.loadAccessToken())
        store.clear()
        assertNull(store.loadAccessToken())
    }
}
